pkg/controller/handlers/mcpcatalog/mcpcatalog.go

```diff
@@ -0,0 +1,467 @@
+						Description: entry.Description,
+						Icon:        entry.Icon,
+						Env:         c.Env,
+						Command:     c.Command,
+						Args:        c.Args,
+						URL:         c.URL,
+						Headers:     c.HTTPHeaders,
+					},
+				}
+			} else if c.URL != "" || c.Remote {
+				if c.URL != "" {
+					if u, err := url.Parse(c.URL); err != nil || u.Hostname() == "localhost" || u.Hostname() == "127.0.0.1" {
+						continue
+					}
+				}
+
+				// Sanitize the headers
+				for i, header := range c.HTTPHeaders {
+					if header.Key == "" {
+						header.Key = header.Name
+					}
+
+					header.Key = strings.ReplaceAll(strings.ToUpper(header.Key), "_", "-")
+
+					c.HTTPHeaders[i] = header
+				}
+
+				addEntry = true
+				catalogEntry.Spec.URLManifest = types.MCPServerCatalogEntryManifest{
+					URL:         entry.URL,
+					GitHubStars: entry.Stars,
+					Metadata:    entry.Metadata,
+					Server: types.MCPServerManifest{
+						Name:        entry.DisplayName,
+						Description: entry.Description,
+						Icon:        entry.Icon,
+						URL:         c.URL,
+						Headers:     c.HTTPHeaders,
+					},
+				}
+			}
+		}
+
+		if addEntry {
+			objs = append(objs, &catalogEntry)
+		}
+	}
+
+	return objs, nil
+}
+
+func (h *Handler) readMCPCatalogDirectory(catalog string) ([]toolreference.CatalogEntryInfo, error) {
+	files, err := os.ReadDir(catalog)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read catalog directory %s: %w", catalog, err)
+	}
+
+	var entries []toolreference.CatalogEntryInfo
+	for _, file := range files {
+		if file.IsDir() {
+			nestedEntries, err := h.readMCPCatalogDirectory(filepath.Join(catalog, file.Name()))
+			if err != nil {
+				return nil, fmt.Errorf("failed to read nested catalog directory %s: %w", file.Name(), err)
+			}
+			entries = append(entries, nestedEntries...)
+		} else {
+			contents, err := os.ReadFile(filepath.Join(catalog, file.Name()))
+			if err != nil {
+				return nil, fmt.Errorf("failed to read catalog file %s: %w", file.Name(), err)
+			}
+
+			var entry toolreference.CatalogEntryInfo
+			if err = json.Unmarshal(contents, &entry); err != nil {
+				return nil, fmt.Errorf("failed to decode catalog file %s: %w", file.Name(), err)
+			}
+			entries = append(entries, entry)
+		}
+	}
+
+	return entries, nil
+}
+
+func isCommandPreferred(existing, newer string) bool {
+	if existing == "" {
+		return true
+	}
+	if newer == "" || existing == "npx" {
+		return false
+	}
+
+	if existing == "uvx" {
+		return newer == "npx"
+	}
+
+	// This would mean that existing is docker and newer is either npx or uvx.
+	return true
+}
+
+// DeleteUnauthorizedMCPServers deletes all MCP servers that are no longer authorized to run.
+// This can happen when a user has launched an MCP server that they used to have permission for,
+// but their access to the catalog that the server came from has been revoked.
+func (h *Handler) DeleteUnauthorizedMCPServers(req router.Request, _ router.Response) error {
```

Comment:
This does seem pretty inefficient.

Can we put the catalog name on the MCP server itself? Then this controller could be on the MCP server object, and it could get its `MCPCatalog` and check to see if it needs to delete itself.

What are your thoughts on this? Is that actually better?

Reply:
I think we should revisit this after we implement what we talked about today, where the admin can create servers directly in a catalog.